### PR TITLE
Capability to download data files from given url

### PR DESCRIPTION
**Pull request by mer-a-o**
This PR adds more capabilities to ```cmake/ecbuild_get_test_data.cmake```. It also adds ```cmake/ecbuild_check_url.cmake``` which will check whether a given URL is available.

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](http://claassistant.ecmwf.int:80/ecmwf/ecbuild?pullRequest=18) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](http://claassistant.ecmwf.int:80/ecmwf/ecbuild?pullRequest=18) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](http://claassistant.ecmwf.int:80/check/ecmwf/ecbuild?pullRequest=18) it.</sub>

**Comment by tlmquintino** [comment]
I agree with @wdeconinck that this should use full URL's instead of NAMES and DIRHOST

**Comment by tlmquintino** [comment]
I don't see the need for 2 macros. I would only provide externally the ecbuild_check_multiurls that I would then rename to ecbuild_check_urls() (note the plural).

File: cmake/ecbuild_check_url.cmake [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      endif()
+
+      # Add to CODE_SUM
+      math( EXPR CODE_SUM "${CODE_SUM} + ${CODE}" )
+    endforeach()
+
+    # Set result
+    set( ${_p_RESULT} ${CODE_SUM} PARENT_SCOPE )
+
+endfunction(ecbuild_check_urls)
```
Comment (wdeconinck) [inline cmake/ecbuild_check_url.cmake:130]: This file should be renamed to match the function name: `ecbuild_check_urls.cmake`

File: cmake/ecbuild_system.cmake [hunk 2]
```diff
@@ -211,6 +211,7 @@ if( NOT ECBUILD_SYSTEM_INITIALISED )
     include( ecbuild_append_to_rpath )
     include( ecbuild_download_resource )
     include( ecbuild_get_test_data )
+    include( ecbuild_check_url )
```
Comment (wdeconinck) [inline cmake/ecbuild_system.cmake:214]:
```suggestion
    include( ecbuild_check_urls )
```
once the file has been renamed

**Comment by mer-a-o** [comment]
@wdeconinck and @tlmquintino thanks for reviewing these changes. I address your comments and all tests pass for me.

File: cmake/ecbuild_check_urls.cmake [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+#
+# Options
+# -------
+#
+# NAMES : required
+#   list of names of the files to check, including the directory structure
+#   on the server hosting test files (if available)
+#
+# RESULT : required
+#   check result (0 if the URL exists, more if not)
```
Comment (oiffrig) [inline cmake/ecbuild_check_urls.cmake:27]:
```suggestion
#   check result (0 if all the URL exist, more if not)
```

File: cmake/ecbuild_check_urls.cmake [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+#
+# Usage
+# -----
+#
+# Check whether files exist on ``<ECBUILD_DOWNLOAD_BASE_URL>/<NAME>``
+# for each name given in the list of ``NAMES``.
+# ``RESULT`` is set to 0 if all the files exist, more if any file is missing.
```
Comment (oiffrig) [inline cmake/ecbuild_check_urls.cmake:34]:
```suggestion
# ``RESULT`` is set to the number of missing files.
```

File: cmake/ecbuild_check_urls.cmake [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      set( ECBUILD_DOWNLOAD_TIMEOUT 30 )
+    endif()
+
+    # Initialise CODE_SUM
+    set( CODE_SUM 0 )
+
+    # Loop over files
+    foreach( NAME ${_p_NAMES} )
+
+      find_program( CURL_PROGRAM curl )
```
Comment (oiffrig) [inline cmake/ecbuild_check_urls.cmake:84]: This should be outside of the loop

File: cmake/ecbuild_check_urls.cmake [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        execute_process(
+          COMMAND ${CURL_PROGRAM} --silent --head --fail --output /dev/null
+                  --retry ${ECBUILD_DOWNLOAD_RETRIES}
+                  --connect-timeout ${ECBUILD_DOWNLOAD_TIMEOUT}
+	          ${ECBUILD_DOWNLOAD_BASE_URL}/${NAME}
+          RESULT_VARIABLE CODE
+        )
+
+      else()
+        find_program( WGET_PROGRAM wget )
```
Comment (oiffrig) [inline cmake/ecbuild_check_urls.cmake:97]: This should be outside of the loop

File: cmake/ecbuild_check_urls.cmake [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            set( WARNING_CANNOT_DOWNLOAD_TEST_DATA 1 CACHE INTERNAL "Couldn't find curl neither wget -- cannot check URL, set result to 0" )
+            mark_as_advanced( WARNING_CANNOT_DOWNLOAD_TEST_DATA )
+          endif()
+
+        endif()
+
+      endif()
+
+      # Add to CODE_SUM
+      math( EXPR CODE_SUM "${CODE_SUM} + ${CODE}" )
```
Comment (oiffrig) [inline cmake/ecbuild_check_urls.cmake:124]:
```suggestion
      if( CODE GREATER 0 )
        math( EXPR CODE_SUM "${CODE_SUM} + 1" )
      endif()
```

**Comment by mer-a-o** [comment]
@oiffrig, thanks for your review. I addressed them in my latest commit.

**Comment by oiffrig** [comment]
Thanks @mer-a-o for the changes. In the current state this does not work if neither curl nor wget is found, and despite being working CMake code, I would prefer not having to use interleaved if / foreach. What I would suggest doing is:
```cmake
find_program( CURL_PROGRAM curl )
if( NOT CURL_PROGRAM )
  find_program( WGET_PROGRAM wget )
  if( NOT WGET_PROGRAM )
    # warn and exit
  endif()
endif()

foreach( NAME ${_p_NAMES} )
  if( CURL_PROGRAM )
    # download with curl
  else()
    # download with wget
  endif()
endforeach()
```

**Pull request merged**